### ISSUE TITLE
Composer 6eb50623e6e00bac4ed893802836e50b62c83434 fails to verify sha1 checksum

### ISSUE THREAD

**Issue by glensc** [composer/composer#4897]
Composer version 1.0-dev (6eb50623e6e00bac4ed893802836e50b62c83434) 2016-02-08 19:03:10

The checksum verification of the file failed (downloaded from https://launchpad.net/php-gettext/trunk/1.0.11/+download/php-gettext-1.0.11.tar.gz)  

i've created gist with the reproducer here (wget and sha1 sum verify included):
https://gist.github.com/glensc/35718a1c48c3fb4066c2

**Comment by glensc** [comment]
known to work with:

Composer version 1.0-dev (72cd6afdfce16f36a9fd786bc1b2f32b851e764f) 2015-12-28 17:35:19

**Comment by curry684** [comment]
I'm not sure I understand your issue with Composer in this. The hash you supplied for an external download appears to be invalid, how is that Composer-related?

**Comment by glensc** [comment]
@curry684 hash is not invalid. please read all information given!

**Comment by glensc** [comment]
i manually walked back to history and found some earlier version giving different error:


```
➔ ~/scm/composer/bin/composer install --no-dev
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Loading composer repositories with package information
Installing dependencies from lock file
  - Installing php-gettext/php-gettext (1.0.11)
    Downloading: 100%         
    Downloading: 100%         
    Downloading: 100%         

                                                                                                                                                                                                                 
  [Composer\Downloader\TransportException]                                                                                                                                                                       
  The "https://launchpad.net/php-gettext/trunk/1.0.11/+download/php-gettext-1.0.11.tar.gz" file could not be downloaded: Peer certificate CN=`launchpadlibrarian.net' did not match expected CN=`launchpad.net'  
  Failed to enable crypto                                                                                                                                                                                        
  failed to open stream: operation failed                                                                                                                                                                        
                     
➔ GIT_DIR=~/scm/composer/.git git log -1
commit 96ff17c520b398d2a102d082bf0d4df3bd5c18a8
Author: Jordi Boggiano <j.boggiano@seld.be>
Date:   Mon Jan 18 18:26:05 2016 +0000

    Cleanups of XDG support, refs #1407
```

**Comment by curry684** [comment]
On current dev-master on Windows it works fine. I stepped through the code in the checks and they match exactly.

On Debian 7.9 it's working fine too:

```
curry@~# composer -V
Composer version 1.0-dev (6eb50623e6e00bac4ed893802836e50b62c83434) 2016-02-08 19:03:10
curry@~# composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing php-gettext/php-gettext (1.0.11)
    Downloading: 100%

Writing lock file
Generating autoload files
curry@~#
```

This suggest it's related to your environment. Can you diff the downloaded object to the wget download?

**Comment by glensc** [comment]
it's broken on at least two different linux machines. with the 6eb50623e6e00bac4ed893802836e50b62c83434 being broken and alpha11 being ok.

the wget download is ok, already shown, but composer download i can not compare as it writes .tar not tar.gz (i.e the cache object).

**Comment by glensc** [comment]
tested php versions:
PHP 5.5.31 (cli) (built: Jan 31 2016 15:52:50) 
PHP 5.5.9-1ubuntu4.14 (cli) (built: Oct 28 2015 01:34:46)

**Comment by glensc** [comment]
i could not fully bisect as some versions between 633e9f9 (good) and b839c7e (bad), give constraint errors:

```
  [UnexpectedValueException]                                                  
  Could not parse version constraint ^1.0.2: Invalid version string "^1.0.2"  
```

**Comment by curry684** [comment]
Are you able to step through Composer code on any of the affected machines? The checksum is validated in `Composer\Download\FileDownloader.php:161`. If you put a breakpoint there you can investigate the file it's trying to hash.

**Comment by glensc** [comment]
[Composer\Download\FileDownloader.php:161](https://github.com/composer/composer/blob/6eb50623e6e00bac4ed893802836e50b62c83434/src/Composer/Downloader/FileDownloader.php#L161)

```php
            if ($checksum && ($v=hash_file('sha1', $fileName)) !== $checksum) {
                throw new \UnexpectedValueException("The [f=$fileName][v=$v][c=$checksum] checksum verification of the file failed (downloaded from ".$url.')');
            }
```

gives:
```
Failed: [UnexpectedValueException] 0: The [f=/home/glen/scm/composer/bug-4897/vendor/php-gettext/php-gettext/f6d706cef269ff1da8121e82d5bbf5d4.gz][v=42aa062d1ec9970f003e9bd3cbfb33e64570423c][c=9e6f402b460ba5e691e6e21f95d23527ab965f71] checksum verification of the file failed (downloaded from https://launchpad.net/php-gettext/trunk/1.0.11/+download/php-gettext-1.0.11.tar.gz)

```

**Comment by glensc** [comment]
my next guess is that download is somehow affected by automatic zlib deflate which composer fails to handle now?

**Comment by curry684** [comment]
But if you compare `/home/glen/scm/composer/bug-4897/vendor/php-gettext/php-gettext/f6d706cef269ff1da8121e82d5bbf5d4.gz` and the wget-downloaded package by hand when breaking/interrupting at line 161, what are their sizes? What happens when you try to `tar xzvf` the interim file manually?

**Comment by glensc** [comment]
well. as said, it's not tar.gz, but .tar:

```
➔ file /home/glen/scm/composer/bug-4897/vendor/php-gettext/php-gettext/271db8bfe02323bdf97e3391b1551113.gz
/home/glen/scm/composer/bug-4897/vendor/php-gettext/php-gettext/271db8bfe02323bdf97e3391b1551113.gz: POSIX tar archive (GNU)
```

composer repacks stuff as tar in it's cache

**Comment by curry684** [comment]
Ping @Seldaek, I don't see what's going wrong here in the code and why.

**Comment by glensc** [comment]
btw, that tmp .gz being moved to .tar was visible in original gist as well:

Writing /home/glen/.composer/cache/files/php-gettext/php-gettext/1.0.11.0-.tar into cache from /home/glen/scm/delfi/dcdb/dcdb-sso/p/vendor/php-gettext/php-gettext/ffdfc3b626d487f8545df71b348c0882.gz

so it seems like checksumming wrong file, but i do not see any relevant changes in `src/Composer/Downloader/FileDownloader.php` at all, i.e comparing with good b056d9a version the diff is just:

```diff
diff --git a/src/Composer/Downloader/FileDownloader.php b/src/Composer/Downloader/FileDownloader.php
index ece495d..77a08cb 100644
--- a/src/Composer/Downloader/FileDownloader.php
+++ b/src/Composer/Downloader/FileDownloader.php
@@ -14,7 +14,6 @@
 
 use Composer\Config;
 use Composer\Cache;
-use Composer\Factory;
 use Composer\IO\IOInterface;
 use Composer\Package\PackageInterface;
 use Composer\Plugin\PluginEvents;
@@ -55,7 +54,7 @@ public function __construct(IOInterface $io, Config $config, EventDispatcher $ev
         $this->io = $io;
         $this->config = $config;
         $this->eventDispatcher = $eventDispatcher;
-        $this->rfs = $rfs ?: Factory::createRemoteFilesystem($this->io, $config);
+        $this->rfs = $rfs ?: new RemoteFilesystem($io, $config);
         $this->filesystem = $filesystem ?: new Filesystem();
         $this->cache = $cache;
 
@@ -141,7 +140,9 @@ protected function doDownload(PackageInterface $package, $path, $url)
                         if ((0 !== $e->getCode() && !in_array($e->getCode(), array(500, 502, 503, 504))) || !$retries) {
                             throw $e;
                         }
-                        $this->io->writeError('    Download failed, retrying...', true, IOInterface::VERBOSE);
+                        if ($this->io->isVerbose()) {
+                            $this->io->writeError('    Download failed, retrying...');
+                        }
                         usleep(500000);
                     }
                 }
```

**Comment by glensc** [comment]
btw, "good" version stores .tar.gz in cache:

```
➔ file /home/glen/.composer/cache/files/php-gettext/php-gettext/1.0.11.0-.tar
/home/glen/.composer/cache/files/php-gettext/php-gettext/1.0.11.0-.tar: gzip compressed data, last modified: Fri Dec 24 23:43:50 2010, from Unix
```

**Comment by glensc** [comment]
maybe this helps:

commit 5d015defb86ca61b09c9d5b9bd042cb231769822 has two parents:
Merge: 939a4aa b056d9a

939a4aa - bad (altho this gives different errors)
b056d9a - good

**Comment by glensc** [comment]
i'm n00b with git bisect, but here's what i obtained:

 2426  2016-02-09 17:00:06 git bisect reset
 2427  2016-02-09 17:00:10 git bisect bad master
 2428  2016-02-09 17:00:25 git bisect good b056d9a 
 2429  2016-02-09 17:00:31 git bisect run ./bug-4897/test.sh


7e30c6782761e6fa33961b4858267fa8fc47247c is the first bad commit
commit 7e30c6782761e6fa33961b4858267fa8fc47247c
Author: Pádraic Brady <padraic.brady@gmail.com>
Date:   Sun Feb 23 10:20:48 2014 +0000

    Add config.json default config option for "disable-tls" (FALSE by default)

:040000 040000 43ba2d10c512b6b7af3e28b50d14449b2395ccad a02a1efaea71cfc13d2714d5235b6ae9d8b5f378 M	src
:040000 040000 ba9d2d415e67a0bb338046a293aa5bff1f4a531d d4901e2663d9c0eb01f34304d408285f25ff681d M	tests
bisect run success


which is weird, as the commit is 2 years old

**Comment by staabm** [comment]
I guess this is related to the recent changes in the archiving

https://github.com/composer/composer/commit/6eb50623e6e00bac4ed893802836e50b62c83434

**Comment by glensc** [comment]
yes, 6eb5062 is broken, but what is it's "parent"?

**Comment by glensc** [comment]
just reverted that PR using patch, and still fails:

```
➔ git checkout master
Your branch is up-to-date with 'origin/master'.
➔ git log -1
commit 6eb50623e6e00bac4ed893802836e50b62c83434
Merge: 9a57de5 c4a9b9b
Author: Jordi Boggiano <j.boggiano@seld.be>
Date:   Mon Feb 8 19:03:10 2016 +0000

    Merge remote-tracking branch 'hfcorriez/compressed_archiver'

➔ curl -siL https://github.com/composer/composer/pull/2706.patch|patch -p1 -R
patching file src/Composer/Package/Archiver/PharArchiver.php
➔ 
```

**Comment by Seldaek** [comment]
Thanks for the repro case, I was able to reproduce under php <5.6 (as hinted by the redirects handled by composer in the output). Nasty little edge case :)

**Comment by glensc** [comment]
actually i reported php versions too:
https://github.com/composer/composer/issues/4897#issuecomment-181876059

but i understand the ticket got little noisy

**Comment by glensc** [comment]
confirming its fixed with 8808638ea993633a6f9e5af240bfcf9714b9201a build
